ipld/read.go

```diff
@@ -18,190 +18,162 @@ import (
-			if err != nil {
-				return nil, err
-			}
+	dataSquare := make([][]byte, edsWidth*edsWidth)
 
-			go sc.retrieveShare(rootCid, true, row, col, dag)
-		}
+	quadrant := pickRandomQuadrant(rowRoots)
+	errGroup, ctx := errgroup.WithContext(parentCtx)
+	errGroup.Go(func() error {
+		return fillQuadrant(ctx, quadrant, dag, true, dataSquare)
+	})
+	if err := errGroup.Wait(); err != nil {
+		return nil, err
 	}
-	// wait until enough data has been collected, too many errors encountered,
-	// or the timeout is reached
-	err := sc.wait()
+	batchAdder := NewNmtNodeAdder(parentCtx, ipld.NewBatch(parentCtx, dag, ipld.MaxSizeBatchOption(batchSize(edsWidth))))
+	tree := wrapper.NewErasuredNamespacedMerkleTree(uint64(edsWidth)/2, nmt.NodeVisitor(batchAdder.Visit))
+	extended, err := rsmt2d.RepairExtendedDataSquare(dah.RowsRoots, dah.ColumnRoots, dataSquare, codec, tree.Constructor)
 	if err != nil {
 		return nil, err
 	}
-	// flatten the square
-	flattened := sc.flatten()
-	// repair the square
-	tree := wrapper.NewErasuredNamespacedMerkleTree(uint64(edsWidth) / 2)
-	return rsmt2d.RepairExtendedDataSquare(rowRoots, colRoots, flattened, codec, tree.Constructor)
-}
 
-// uniqueRandNumbers generates count unique random numbers with a max of max
-func uniqueRandNumbers(count, max int) []uint32 {
-	if count > max {
-		panic(fmt.Sprintf("cannot create %d unique samples from a max of %d", count, max))
-	}
-	samples := make(map[uint32]struct{}, count)
-	for i := 0; i < count; {
-		// nolint:gosec // G404: Use of weak random number generator
-		sample := uint32(rand.Intn(max))
-		if _, has := samples[sample]; has {
-			continue
-		}
-		samples[sample] = struct{}{}
-		i++
-	}
-	out := make([]uint32, count)
-	counter := 0
-	for s := range samples {
-		out[counter] = s
-		counter++
-	}
-	return out
+	return extended, batchAdder.Commit()
 }
 
-type index struct {
-	row uint32
-	col uint32
+type quadrant struct {
+	isLeftSubtree bool
+	from          int
+	chunk         [][]byte //roots
 }
 
-type indexedShare struct {
-	data []byte
-	index
-}
+func pickRandomQuadrant(roots [][]byte) *quadrant {
+	edsWidth := len(roots)
+	// get the random number from 1 to 4.
+	q := rand.Intn(4) + 1
+	quadrant := &quadrant{chunk: make([][]byte, edsWidth/2)}
+	// quadrants 1 and 3 corresponds to left subtree,
+	// 2 and 4 to the right subtree
+	/* | 1 | 2 |
+	   | 3 | 4 |
+	*/
+	// choose sub-tree
+	if q%2 == 1 {
+		quadrant.isLeftSubtree = true
+	}
 
-// shareCounter is a thread safe tallying mechanism for share retrieval
-type shareCounter struct {
-	// all shares
-	shares map[index][]byte
-	// number of shares successfully collected
-	counter uint32
-	// the width of the extended data square
-	edsWidth uint32
-	// the minimum shares needed to repair the extended data square
-	minSharesNeeded uint32
+	// define range of shares for sampling
+	if q > 2 {
+		quadrant.from = edsWidth / 2
+		copy(quadrant.chunk, roots[edsWidth/2:])
```

Comment:
You could also point to the full root slice in the code where you use roots, instead of copying the half.

Reply:
But that is an optional nit

Reply:
changed type to `cid.Cid`. So, there is no needs of copying anymore.